File: answers/Javascript/@CFCIfe/14. Number in words.js
function numToWords() {
  let user = prompt("Enter a number: "),
    userNum = Number(user),
    arrOfNumber = user.split(""),
    unit = arrOfNumber.slice(-1),
    tens = arrOfNumber.slice(-2, -1),
    hundreds = arrOfNumber.slice(-3, -2),
    thousands = arrOfNumber.slice(-5, -3);
  let unitNames = [
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
    "ten",
    "eleven",
    "twelve",
    "thirteen",
    "fourteen",
    "fifteen",
    "sixteen",
    "seventeen",
    "eighteen",
    "nineteen"
  ];
  let tensNames = [
    "",
    "twenty",
    "thirty",
    "forty",
    "fifty",
    "sixty",
    "seventy",
    "eighty",
    "ninety"
  ];

  if ((arrOfNumber.length === 1 || arrOfNumber.length === 2) && userNum < 20) {
    return unitNames[userNum - 1];
  } else if (
    (arrOfNumber.length === 1 || arrOfNumber.length === 2) &&
    userNum >= 20 &&
    unit == 0
  ) {
    return tensNames[tens - 1];
  } else if (
    (arrOfNumber.length === 1 || arrOfNumber.length === 2) &&
    userNum >= 20 &&
    unit != 0
  ) {
    return `${tensNames[tens - 1]}-${unitNames[unit - 1]}`;
  } else if (arrOfNumber.length === 3) {
    return `${unitNames[hundreds - 1] +
      " hundred" +
      (tens != 0 ? " and " + tensNames[tens - 1] : "") +
      " " +
      (unit != 0 ? unitNames[unit - 1] : "")}`;
  } else if (arrOfNumber.length === 4 || arrOfNumber.length === 5) {
    return `${(thousands.join("").length <= 2 && thousands.join("") < 20
      ? unitNames[thousands.join("") - 1]
      : tensNames[thousands[0] - 1] + " " + unitNames[thousands[1] - 1]) +
      " thousand, " +
      (hundreds != 0 ? unitNames[hundreds - 1] + " hundred" : "") +
      (tens != 0 ? " and " + tensNames[tens - 1] : "") +
      " " +
      (unit != 0 ? unitNames[unit - 1] : "")}`;
  } else {
    return null;
  }
}

File: answers/Python/@oseme-techguy/14-number-in-words.py
"""
    Solution to Numbers in Words
    Based off this Algorithm: https://stackoverflow.com/a/3299672/7396801
"""

import math

zero_to_nineteen_map = [
    'zero', 'one', 'two', 'three', 'four', 'five', 'six',
    'seven', 'eight', 'nine', 'ten', 'eleven', 'twelve', 'thirteen',
    'fourteen', 'fifteen', 'sixteen', 'seventeen', 'eighteen', 'nineteen'
]

tens_map = [
    'twenty', 'thirty', 'forty', 'fifty', 'sixty', 'seventy', 'eighty', 'ninety'
]

denom_map = [
    '' ,'thousand', 'million', 'billion', 'trillion', 'quadrillion',
    'quintillion', 'sextillion', 'septillion', 'octillion', 'nonillion',
    'decillion', 'undecillion', 'duodecillion', 'tredecillion', 'quattuordecillion',
    'sexdecillion', 'septendecillion', 'octodecillion', 'novemdecillion', 'vigintillion',
]

def hundred_to_words(number):
    if (number < 20):
        return zero_to_nineteen_map[number]
    for index,value in enumerate(tens_map):
        dval = 20 + 10 * index
        if (dval + 10 > number):
            if ((number % 10) != 0):
                return value + "-" + zero_to_nineteen_map[number % 10]
            return value


def thousand_to_words(number):
    word = ''
    rem = math.floor(number / 100)
    mod = number % 100
    if rem > 0:
        word = zero_to_nineteen_map[rem] + ' hundred'
        if mod > 0:
            word = word + ' '
    if mod > 0:
        word = word + hundred_to_words(mod)
    return word


def number_to_words(number):
    if number < 100:
        return hundred_to_words(number)

    if number < 1000:
        return thousand_to_words(number)

    for index,value in enumerate(denom_map):
        didx = index - 1
        dval = math.floor(float(math.pow(1000, index)))

        if (dval > number):
            mod = int(float(math.pow(1000, didx)))
            
            l = math.floor(number / mod)
            r = number - (l * mod)
            ret = thousand_to_words(l) + ' ' + denom_map[didx]
            if r > 0:
                ret = ret + ', ' + number_to_words(r)
            return ret


if __name__ == '__main__':

    while True:
        number = input('Enter a number: ')

        # type checking here
        try:
            number = int(number)
        except ValueError:
            number = -1

        if number < 0: # corner case
            print('Invalid Amount')
            exit(0)

        english_number = number_to_words(number)
        print(english_number + '\n')

